### DISCUSSION TITLE
Add fullnameof() similar to nameof()

### DISCUSSION THREAD

**Discussion by joshmouch** [dotnet/csharplang#6828] [Language Ideas] [answered]
I use "nameof()" often for use in constants, strings, logging, etc.  It's a very convenient way to keep values and documentation up-to-date when using common Visual Studio refactoring tools.   There are infinite examples, but here's one:

```
public static class ConstantCollection {
  public const string Prefix = $"{nameof(ConstantCollection)}";

  public static class ConstantCategory {
    public const string Prefix = $"{ConstantCollection.Prefix}.{nameof(ConstantCategory)}";

    public static const Entity1 {
      public const string Prefix = $"{ConstantCategory.Prefix}.{nameof(Entity1)}";

      public const string Value1 = $"{Prefix}.{nameof(Value1)}";
    }
  }
}
```

It would be incredibly useful to also allow for the fully qualified type name to be used compile-time.  Something like `fullname(Entity1)` would return `"ConstantCollection.Category.Entity1.Value1"`.

In this StackOverflow, it was said that this has been discussed in the past, and the Type.GetFullName() API was considered good enough.  I would disagree because the power of "name()" is that it can be used during compile-time.

https://stackoverflow.com/questions/40855818/how-to-use-nameof-to-get-the-fully-qualified-name-of-a-property-in-a-class-in-c

**Comment by svick** [comment, marked as answer]
This is a duplicate of https://github.com/dotnet/csharplang/discussions/701.

**Comment by jnm2** [comment]
I'm proposing making `typeof(`...`).FullName` a constant, which will help with some of these use cases: https://github.com/dotnet/csharplang/issues/8505

It would be great to hear in that thread about real-world use cases that support doing this, as well as real-world use cases for `fullnameof` that _aren't_ addressed by that proposal and don't have a good solution today.
